File: 25_2/s_5/exo_6/exo_6.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <iomanip>

using namespace std;

class Node {
public:
    double weight;
    vector<Node*> children;

    Node(double w) : weight(w) {}
};


Node* buildTree(int depth, double w, int n) {
    Node* root = new Node(w);

    if (depth == 0)
        return root;

    for (int i = 0; i < n; i++)
        root->children.push_back(buildTree(depth - 1, w / n, n));

    return root;
}

// DFS суммирование
double dfsSum(Node* root) {
    double sum = root->weight;
    for (auto child : root->children)
        sum += dfsSum(child);
    return sum;
}

// BFS суммирование
double bfsSum(Node* root) {
    queue<Node*> q;
    q.push(root);
    double sum = 0;

    while (!q.empty()) {
        Node* current = q.front(); q.pop();
        sum += current->weight;
        for (auto child : current->children)
            q.push(child);
    }
    return sum;
}

// DFS с инвертированием весов
double dfsFlip(Node* root) {
    root->weight = -root->weight;
    double sum = root->weight;
    for (auto child : root->children)
        sum += dfsFlip(child);
    return sum;
}

// BFS с инвертированием весов
double bfsFlip(Node* root) {
    queue<Node*> q;
    q.push(root);
    double sum = 0;

    while (!q.empty()) {
        Node* current = q.front(); q.pop();
        current->weight = -current->weight;
        sum += current->weight;
        for (auto child : current->children)
            q.push(child);
    }
    return sum;
}

// Рекурсивный BFS (уровень за уровнем)
double bfsRecursive(vector<Node*> level) {
    if (level.empty())
        return 0;

    vector<Node*> nextLevel;
    double sum = 0;

    for (auto node : level) {
        sum += node->weight;
        for (auto child : node->children)
            nextLevel.push_back(child);
    }

    return sum + bfsRecursive(nextLevel);
}


void printTree(Node* node, string prefix = "", bool isLast = true) {
   
    cout << prefix;

    cout << (isLast ? "└── " : "├── ") << node->weight << endl;

   
    string childPrefix = prefix + (isLast ? "    " : "│   ");

    for (size_t i = 0; i < node->children.size(); i++) {
        printTree(node->children[i], childPrefix, i == node->children.size() - 1);
    }
}

int main() {
    int n = 5;
    int depth = 3;
    double rootWeight = 1.0 / (depth + 1);

    Node* root = buildTree(depth, rootWeight, n);

    cout << fixed << setprecision(8);

    cout << "\nTree structure:\n";
    printTree(root);

    cout << "\n---------------------------------\n";
    cout << "DFS sum: " << dfsSum(root) << endl;
    cout << "BFS sum: " << bfsSum(root) << endl;

    cout << "\nDFS flip #1: " << dfsFlip(root) << endl;
    cout << "BFS flip #1: " << bfsFlip(root) << endl;

    cout << "\nDFS flip #2: " << dfsFlip(root) << endl;
    cout << "BFS flip #2: " << bfsFlip(root) << endl;

    cout << "\nRecursive BFS sum: " << bfsRecursive({root}) << endl;
    cout << "\n---------------------------------\n";

    return 0;
}

File: 25_2/s_2/exo_6/exo_6.py
class Node:
    def __init__(self, weight):
        self.weight = weight
        self.children = []


def build_tree(depth, weight, n):
    node = Node(weight)

    if depth == 0:
        return node

    for _ in range(n):
        child = build_tree(depth - 1, weight / n, n)
        node.children.append(child)

    return node


def dfs_sum(node):
    total = node.weight
    for child in node.children:
        total += dfs_sum(child)
    return total


def bfs_sum(root):
    queue = [root]
    total = 0.0

    while queue:
        current = queue.pop(0)
        total += current.weight

        for child in current.children:
            queue.append(child)

    return total


def dfs_flip(node):
    node.weight = -node.weight
    total = node.weight

    for child in node.children:
        total += dfs_flip(child)

    return total


def bfs_flip(root):
    queue = [root]
    total = 0.0

    while queue:
        current = queue.pop(0)
        current.weight = -current.weight
        total += current.weight

        for child in current.children:
            queue.append(child)

    return total


def bfs_recursive(level_nodes):
    if not level_nodes:
        return 0.0

    next_level = []
    level_sum = 0.0

    for node in level_nodes:
        level_sum += node.weight
        for child in node.children:
            next_level.append(child)

    return level_sum + bfs_recursive(next_level)


def print_tree(node, level=0):
    indent = "   " * level
    print(f"{indent}weight = {node.weight:.6f}")
    for child in node.children:
        print_tree(child, level + 1)


if __name__ == "__main__":
    n = 5
    depth = 3

    # starting weight for the root
    root_weight = 1.0 / (depth + 1)

    root = build_tree(depth, root_weight, n)

    print("Tree Structure (basic print)")
    print_tree(root)

    print("\nSum Checks")
    print("DFS sum:", dfs_sum(root))
    print("BFS sum:", bfs_sum(root))

    print("\nFirst Flip")
    print("DFS flip:", dfs_flip(root))
    print("BFS flip:", bfs_flip(root))

    print("\nSecond Flip")
    print("DFS flip #2:", dfs_flip(root))
    print("BFS flip #2:", bfs_flip(root))

    print("\nRecursive BFS")
    print("Recursive BFS sum:", bfs_recursive([root]))

    print("\nProgram Finished")
